Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp

```diff
@@ -32,311 +32,259 @@
-            continue;
+        if (!equalLettersIgnoringASCIICase(contentType->mimeType(), "multipart/x-mixed-replace"_s))
+            return std::nullopt;
 
-        auto boundary = extractBoundaryFromContentType(contentType);
-        if (!boundary)
-            continue;
+        auto boundary = contentType->parameterValueForName("boundary"_s);
+        if (boundary.isEmpty())
+            return std::nullopt;
 
-        return String("--" + *boundary);
+        return makeString("--", boundary).latin1();
     }
 
     return std::nullopt;
 }
 
-std::optional<String> CurlMultipartHandle::extractBoundaryFromContentType(const String& contentType)
+std::unique_ptr<CurlMultipartHandle> CurlMultipartHandle::createIfNeeded(CurlMultipartHandleClient& client, const CurlResponse& response)
 {
-    static const size_t length = strlen("boundary=");
-
-    auto boundaryStart = contentType.findIgnoringASCIICase("boundary="_s);
-    if (boundaryStart == notFound)
-        return std::nullopt;
-
-    boundaryStart += length;
-    size_t boundaryEnd = 0;
-
-    if (contentType[boundaryStart] == '"') {
-        // Boundary value starts with a " quote. Search for the closing one.
-        ++boundaryStart;
-        boundaryEnd = contentType.find('"', boundaryStart);
-        if (boundaryEnd == notFound)
-            return std::nullopt;
-    } else if (contentType[boundaryStart] == '\'') {
-        // Boundary value starts with a ' quote. Search for the closing one.
-        ++boundaryStart;
-        boundaryEnd = contentType.find('\'', boundaryStart);
-        if (boundaryEnd == notFound)
-            return std::nullopt;
-    } else {
-        // Check for the end of the boundary. That can be a semicolon or a newline.
-        boundaryEnd = contentType.find(';', boundaryStart);
-        if (boundaryEnd == notFound)
-            boundaryEnd = contentType.length();
-    }
-
-    // The boundary end should not be before the start
-    if (boundaryEnd <= boundaryStart)
-        return std::nullopt;
+    auto boundary = extractBoundary(response);
+    if (!boundary)
+        return nullptr;
 
-    return contentType.substring(boundaryStart, boundaryEnd - boundaryStart);
+    return makeUnique<CurlMultipartHandle>(client, WTFMove(*boundary));
 }
 
-CurlMultipartHandle::CurlMultipartHandle(CurlMultipartHandleClient& client, const String& boundary)
+CurlMultipartHandle::CurlMultipartHandle(CurlMultipartHandleClient& client, CString&& boundary)
     : m_client(client)
-    , m_boundary(boundary)
+    , m_boundary(WTFMove(boundary))
 {
-
 }
 
-void CurlMultipartHandle::didReceiveData(const SharedBuffer& buffer)
+void CurlMultipartHandle::didReceiveMessage(const SharedBuffer& buffer)
 {
-    if (m_state == State::End)
+    if (m_state == State::WaitingForTerminate || m_state == State::End || m_didCompleteMessage)
         return; // The handler is closed down so ignore everything.
 
     m_buffer.append(buffer.data(), buffer.size());
```

Comment:
Do you need `m_buffer.makeContiguous();` here?

Reply:
> Is the buffer isContiguous()?

No, it isn't. And we don't need `buffer.makeContiguous();`.

`m_contiguous` is `true` only when generated using the following four constructors.

- SharedBuffer::SharedBuffer()
- SharedBuffer::SharedBuffer(Ref<const DataSegment>&& segment)
- SharedBuffer::SharedBuffer(Ref<FragmentedSharedBuffer>&& contiguousBuffer)
- SharedBuffer::SharedBuffer(FileSystem::MappedFileData&& data)

https://github.com/WebKit/WebKit/blob/3fa12a1b0c5d4be171d8824ad8fd765ffd22ac57/Source/WebCore/platform/SharedBuffer.cpp#L447-L472

On the other hand, the `buffer` was created using `SharedBuffer::create(const uint8_t*, size_t);`, so we do not use the above constructor. Therefore, `m_contiguous` remains false. In addition, there is no need to call `makeContiguous` because no data has been added or modified to this SharedBuffer since it was created.

https://github.com/WebKit/WebKit/blob/3fa12a1b0c5d4be171d8824ad8fd765ffd22ac57/Source/WebCore/platform/network/curl/CurlRequest.cpp#L624

https://github.com/WebKit/WebKit/blob/3fa12a1b0c5d4be171d8824ad8fd765ffd22ac57/Source/WebCore/platform/SharedBuffer.cpp#L45-L48

https://github.com/WebKit/WebKit/blob/3fa12a1b0c5d4be171d8824ad8fd765ffd22ac57/Source/WebCore/platform/SharedBuffer.h#L221

Reply:
Then, using `Vector<uint8_t>` seems to be better to pass the data. Anyway, it's another problem.